cpp/validate_harvested_records.cc

```diff
@@ -322,58 +250,70 @@ void SendEmail(const std::string &email_address, const std::string &message_subj
-                   unsigned * const new_record_count, std::string * const error_message)
+bool RecordIsValid(const MARC::Record &record, const std::vector<const FieldValidator *> &regular_article_field_validators,
+                   const std::vector<const FieldValidator *> &review_article_field_validators,
+                   std::string * const reasons_for_being_invalid)
 {
-    const std::string zeder_id(record.getFirstSubfieldValue("ZID", 'a'));
-    const std::string zeder_instance(record.getFirstSubfieldValue("ZID", 'b'));
-    if (zeder_id.empty() or zeder_instance.empty())
-        LOG_ERROR("Record w/ control number \"" + record.getControlNumber() + "\" has either no zeder_id or no zeder_instance!");
-
-    const auto journal_name(record.getSuperiorTitle());
-    if (journal_name.empty()) {
-        *error_message = "Record w/ control number \"" + record.getControlNumber() + "\" is missing a superior title!";
-        return false;
-    }
-
-    auto journal_name_and_info(journal_name_to_info_map->find(journal_name));
-    bool first_record(false); // True if the current record is the first encounter of a journal
-    if (journal_name_and_info == journal_name_to_info_map->end()) {
-        first_record = true;
-        JournalInfo new_journal_info;
-        LoadFromDatabaseOrCreateFromScratch(db_connection, zeder_id, zeder_instance, &new_journal_info);
-        (*journal_name_to_info_map)[journal_name] = new_journal_info;
-        journal_name_and_info = journal_name_to_info_map->find(journal_name);
+    reasons_for_being_invalid->clear();
+
+    const auto zid_field(record.findTag("ZID"));
+    if (unlikely(zid_field == record.end()))
+        LOG_ERROR("record is missing a ZID field!");
+    else if (zid_field->getFirstSubfieldWithCode('a').empty())
+        LOG_ERROR("record is missing an a-subfield in the existing ZID field!");
+    const auto ZID(zid_field->getFirstSubfieldWithCode('a'));
+
+    // 0. Check that requirements for all records, independent of type or journal are met:
+    CheckGenericRequirements(record, reasons_for_being_invalid);
+
+    // 1. Check that present fields meet all the requirements:
+    MARC::Tag last_tag("   ");
+    const auto &field_validators(record.isReviewArticle() ? review_article_field_validators : regular_article_field_validators);
+    std::set<std::string> present_tags;
+    for (const auto &field : record) {
+        const auto current_tag(field.getTag());
+        if (current_tag == last_tag and not field.isRepeatableField())
+            reasons_for_being_invalid->append(current_tag.toString() + " is not a repeatable field\n");
+        last_tag = current_tag;
+        present_tags.emplace(current_tag.toString());
+
+        for (const auto field_validator : field_validators) {
+            std::string reason_for_being_invalid;
+            if (field_validator->foundRuleMatch(ZID, field, &reason_for_being_invalid)) {
```

Comment:
I guess this should not only be performed for every field, but for every single subfield.

Reply:
No! (Please talk to me about why.)

Reply:
As discussed, if we want to keep the behaviour like this, we should definitely speak to @socheres about it before merging.